openhtf/io/station_api.py

```diff
@@ -0,0 +1,574 @@
+    # together, you could get Test UID collisions.
+    self._history = history.History()
+    # Maps test UID to RemoteTest, so we can reuse old RemoteTest objects in
+    # order to benefit from their saved state (phases and logs).
+    self._known_tests = {}
+    # Shared proxy and lock used for synchronous calls we expect to be fast.
+    # Long-polling calls should use the 'proxy' attribute directly instead, as
+    # it creates a new ServerProxy object each time, avoiding the danger of
+    # blocking short requests with long-running ones.
+    self._shared_proxy = self.make_proxy()
+
+  def __hash__(self):
+    return hash(self._station_info)
+
+  def __eq__(self, other):
+    return (type(self) == type(other) and
+            self._station_info == other._station_info)
+
+  def __ne__(self, other):
+    return not self.__eq__(other)
+
+  def make_proxy(self):
+    """Make a new ServerProxy for this station."""
+    proxy = xmlrpclib.ServerProxy(
+        'http://%s:%s' % (self.host, self.station_api_port),
+        allow_none=True)
+    proxy._lock = threading.Lock()
+    return proxy
+
+  def __str__(self):
+    return 'Station %s@%s:%s, Listening on %s' % (
+        self.station_id, self.host, self.station_api_port,
+        self.station_api_bind_address)
+
+  def __getattr__(self, attr):
+    return getattr(self._station_info, attr)
+
+  @classmethod
+  def discover_stations(cls, timeout_s=3):
+    """Discover Stations, yielding them as they're found."""
+    for host, response in multicast.send(conf.station_discovery_string,
+                                         timeout_s=timeout_s,
+                                         **MULTICAST_KWARGS()):
+      try:
+        yield cls(StationInfo(host, **json.loads(response)))
+      except ValueError:
+        _LOG.debug('Received malformed JSON from %s: %s', host, response)
+      except TypeError:
+        _LOG.debug('Received invalid discovery response from %s: %s',
+                   host, response, exc_info=True)
+
+  @property
+  def tests(self):
+    """List active Test instances on this station."""
+    print('listing tests')
+    with self._shared_proxy._lock:
+      print('acquired lock')
+      tests = self._shared_proxy.list_tests()
+    print('received %s tests' % len(tests))
+
+    updated_tests = {}
+    for test_dict in tests:
+      test_uid = test_dict['test_uid']
+      ctime_millis = test_dict['created_time_millis']
+      if (test_uid not in self._known_tests or
+          ctime_millis != self._known_tests[test_uid].created_time_millis):
+        updated_tests[test_uid] = RemoteTest(
+            self.make_proxy, self._shared_proxy, self._history, **test_dict)
+      else:
+        updated_tests[test_uid] = self._known_tests[test_uid]
+
+    self._known_tests = updated_tests
+    return self._known_tests.itervalues()
+
+
+### Server-side objects below here. ###
+
+
+class StationApi(object):
+
+  UID = '%s:%s' % (os.getpid(), util.TimeMillis())
```

Comment:
thought about this when doing it, but i like that the UID gives a knowledgeable sysadmin some traceability (ie, you can look up the PID), and the randomness doesn't really buy anything since PID is guaranteed unique within a millisecond anyway (you've got bigger problems if you have cycled through all PIDs in <1ms).